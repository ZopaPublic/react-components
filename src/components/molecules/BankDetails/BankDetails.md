# Summary

`<BankDetails />` renders the bank detail with option to copy value to clipboard using [react-use-clipboard](https://www.npmjs.com/package/react-use-clipboard)

## Example

```tsx
import { BankDetails, Text } from '@zopauk/react-components';

const items = [
  {
    label: 'Payee',
    text: 'Chris Jeffries',
    copyText: 'Chris Jeffries',
  },
  {
    label: 'Sort code',
    text: '00-00-00',
    copyText: '000000',
  },
  {
    label: 'Account number',
    text: '12345678',
    copyText: '12345678',
  },
  {
    label: 'Reference',
    text: 'Anything you want!',
  },
];

<div>
  {items.map((bankDetail, index) => {
    return (
      <BankDetails key={`bank-detail-${index}`} copyText={bankDetail.copyText}>
        <Text size="lead" as="p" weight="semiBold">
          {bankDetail.text}
        </Text>
        <Text size="small" color={colors.grey}>
          {bankDetail.label}
        </Text>
      </BankDetails>
    );
  })}
</div>;
```
