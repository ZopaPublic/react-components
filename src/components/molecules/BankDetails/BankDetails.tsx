import React, { HTMLAttributes } from 'react';
import styled from 'styled-components';
import useClipboard from 'react-use-clipboard';
import classnames from 'classnames';

import { colors } from '../../../constants';
import FlexCol from '../../layout/FlexCol/FlexCol';
import FlexRow from '../../layout/FlexRow/FlexRow';
import Button from '../../atoms/Button/Button';

interface BankDetailsProps extends HTMLAttributes<HTMLDivElement> {
  /**Set to null to not render a copy Button */
  copyText?: string;
}

const ItemWrapper = styled.div`
  border: 1px solid ${colors.greyLighter};

  &:first-of-type {
    border-top-left-radius: 8px;
    border-top-right-radius: 8px;
  }

  &:last-of-type {
    border-bottom-left-radius: 8px;
    border-bottom-right-radius: 8px;
    box-shadow: 0 1px 0 0 ${colors.greyLight};
  }
`;

const CopyText = styled(Button)`
  cursor: pointer;
  padding: 8px;

  &:hover,
  &:focus {
    background: transparent;
    color: ${colors.actionDark};
  }
`;

const BankDetails = ({ children, copyText = '', className, ...props }: BankDetailsProps) => {
  const [isCopied, setCopied] = useClipboard(copyText, {
    successDuration: 2000,
  });

  return (
    <ItemWrapper {...props} className={classnames('pl-4 pr-2 py-5', className || '')}>
      <FlexRow justify="space-between" align="center">
        <FlexCol xs="fill">{children}</FlexCol>
        {copyText && (
          <FlexCol xs="auto">
            <CopyText
              styling="link"
              onClick={() => {
                if (isCopied) {
                  return;
                }
                setCopied();
              }}
            >
              {isCopied ? 'Copied' : 'Copy'}
            </CopyText>
          </FlexCol>
        )}
      </FlexRow>
    </ItemWrapper>
  );
};

export default BankDetails;
