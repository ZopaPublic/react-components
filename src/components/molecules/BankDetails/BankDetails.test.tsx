import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import 'jest-styled-components';

import BankDetails from './BankDetails';
import axe from '../../../../axe-helper';

const mockFn = jest.fn();
let mockIsCopied = false;

jest.mock('react-use-clipboard', () => {
  return () => [mockIsCopied, mockFn];
});

describe('<BankDetails />', () => {
  it('renders the fallback option', () => {
    render(<BankDetails copyText={'123456789'}>Test</BankDetails>);
    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(screen.getByText('Copy')).toBeInTheDocument();
  });

  it('renders the fallback option', () => {
    render(<BankDetails>Test</BankDetails>);
    expect(screen.getByText('Test')).toBeInTheDocument();
    expect(screen.queryByText('Copy')).not.toBeInTheDocument();
  });

  it('should call setCopied when copy button is clicked', () => {
    render(<BankDetails copyText={'123456789'}>Test</BankDetails>);
    act(() => {
      fireEvent.click(screen.getByText('Copy'));
    });
    expect(mockFn).toHaveBeenCalledTimes(1);
  });

  it('should disable the button when isCopied is true', () => {
    mockIsCopied = true;
    render(<BankDetails copyText={'123456789'}>Test</BankDetails>);
    expect(screen.getByText('Copied')).toBeInTheDocument();
  });

  it('renders the component with no props and no a11y violations', async () => {
    jest.useRealTimers();
    const { container } = render(
      <BankDetails copyText="1234" className="mb-4">
        12-34
      </BankDetails>,
    );
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });
});
