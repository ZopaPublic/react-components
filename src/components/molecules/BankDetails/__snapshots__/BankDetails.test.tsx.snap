// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BankDetails /> renders the component with no props and no a11y violations 1`] = `
.c2 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

.c3 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-right: -12px;
  margin-left: -12px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  outline: 0;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 12px 24px;
  background: transparent;
  border-radius: 8px;
  border: 1px solid transparent;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  color: #21247F;
  font-size: 16px;
  font-weight: 600;
  line-height: 24px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c4:focus:not(:active) {
  border: 1px solid #FFFFFF;
  box-shadow: 0 0 4px #3B46C4;
}

.c4:active {
  opacity: 0.8;
  box-shadow: unset;
  border: 1px solid transparent;
}

.c4:hover {
  background: #EAEBFA;
}

.c0 {
  border: 1px solid #EFEFEF;
}

.c0:first-of-type {
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
}

.c0:last-of-type {
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
  box-shadow: 0 1px 0 0 #D4D7D9;
}

.c5 {
  cursor: pointer;
  padding: 8px;
}

.c5:hover,
.c5:focus {
  background: transparent;
  color: #21247F;
}

@media (min-width:0px) {
  .c2 {
    display: block;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    max-width: 100%;
  }
}

@media (min-width:0px) {
  .c3 {
    display: block;
    -webkit-flex: 0 0 auto;
    -ms-flex: 0 0 auto;
    flex: 0 0 auto;
    width: auto;
    max-width: none;
  }
}

<div>
  <div
    class="c0 pl-4 pr-2 py-5 mb-4"
  >
    <div
      class="c1"
      cols="12"
      direction="row"
    >
      <div
        class="c2"
        cols="12"
      >
        12-34
      </div>
      <div
        class="c3"
        cols="12"
      >
        <button
          class="c4 c5"
          styling="link"
          theme="[object Object]"
        >
          Copied
        </button>
      </div>
    </div>
  </div>
</div>
`;
